### DISCUSSION TITLE
deformation induced grain nucleation for dynamic recrystallization

### DISCUSSION THREAD

**Discussion by jbair34** [idaholab/moose#20093] [Q&A Modules: Phase field] [answered]
Hi All,
I am having trouble figuring out how to make new grains nucleate in a dynamic recrystallization model with deformation induced nucleation. I am trying to follow the https://mooseframework.inl.gov/modules/phase_field/Nucleation/DiscreteNucleation.html instructions below:

"Direct order parameter modification
...
Nucleation of new grains in a polycrystalline model with GrainTracker requires the use of reserved (or staging) order parameters (order parameters which the grain tracker detects new grains in, but never uses as a target for remapping) in which new grain nuclei can be inserted and successively get picked up and remapped onto the evolving order parameters.

Apply both a DiscreteNucleationForce and a Reaction kernel to a reserved order parameter and set the hold_time in the inserter to 0. This will cause the modification of the reserved OP during only a single timestep, which will be enough for the GrainTracker to pick up and remap the new nucleus."

From that I have made a discrete nucleation inserter which has a rate dependent on the stress as a function of dislocation density and it does make the nuclei as the domain deforms. However, the nuclei just disappear without forming new grains. Is there something missing in this description where I can set a new random euler angle in the nucleus or make sure the relevant non conserved OP goes to 0 or something so a new grain will form?

I have combined moose/modules/combined/examples/phase_field-mechanics/hex_grain_growth_2D_eldrforce.i with the nucleation inputs in moose/modules/phase_field/examples/nucleation/refine.i which I modified as a test to have a probability linked to the vonmises stress. The probability seems to work fine putting nuclei in reasonable locations they just don't seem to stick...
Thanks,
Jake

**Comment by laagesen** [comment, marked as answer]
Hi Jake- when you say the particles are appearing and disappearing, are you saying that based on what's happening in the reserved order parameter? Or are you looking an AuxVariable generated by DiscreteNucleationAux? I'm trying to figure out if the problem is with the reserved OP not getting modified, or with the GrainTracker not picking it up and remapping it.

**Reply by jbair34** [reply]
I am looking at the reserved order parameter, I can see where the nuclei are supposed to be, but the maximum value of my reserved order parameter (gr7 in my case) in the center of the nuclei is around 0.06 which doesn't seem like enough to cause remapping to me. I set the nuc_value to 1 in the discreteNucleationForce kernel but that doesn't seem to be happening. I wasn't paying attention to how low they were before my last post so sorry for missing that information the first time around.

**Reply by laagesen** [reply]
Are you sure you have the reaction kernel in place and also assigned to the reserved OP? There should only be 2 kernels associated with the reserved OP, DiscreteNucleationForce and Reaction. I thought the Reaction kernel should make the reserved OP value to match the DiscreteNucleationForce value exactly.

**Reply by jbair34** [reply]
Here is my input for the kernels (reserved OP 3rd and 4th entries):

```
[Kernels]
  [./PolycrystalKernel]
  [../]
  [./PolycrystalElasticDrivingForce]
  [../]
  [./nuc_force]
    type = DiscreteNucleationForce
    variable = gr7
    map = map
    no_nucleus_value = 0
    nucleus_value = 2
  [../]
  [./nuc_react]
    type = Reaction
    variable = gr7
  [../]
  [./TensorMechanics]
    displacements = 'disp_x disp_y'
  [../]
[]
```

My grain tracker set up for 8 OPs is as follows:

```
[./grain_tracker]
    type = GrainTrackerElasticity
    threshold = 0.2
    compute_var_to_feature_map = true
    execute_on = 'initial timestep_begin'
    flood_entity_type = ELEMENTAL

    fill_method = symmetric9
    C_ijkl = '1.27e5 0.708e5 0.708e5 1.27e5 0.708e5 1.27e5 0.7355e5 0.7355e5 0.7355e5'
    euler_angle_provider = euler_angle_file
    reserve_op = 1
    reserve_op_threshold = 0.95
  [../]
```

I double checked and it does appear the gr7 is the one that is reserved. I don't see an option to set a value in the Reaction Kernel options...

**Reply by laagesen** [reply]
Are you sure that the PolycrystalKernel action isn't populating kernels for gr7? If you are setting a global num_ops parameter to 8 it is probably adding the kernels for gr7. It might be worth using the DumpObjectsProblem syntax to see exactly what kernels are getting added to your input file. To me it seems like something is keeping gr7 from getting set to exactly the value specified by DiscreteNucleationForce.

https://mooseframework.inl.gov/source/problems/DumpObjectsProblem.html

**Reply by jbair34** [reply]
I am having trouble getting the DumpObjectsProblem to actually output anything besides Solve Converged on my input file. I can get it to reproduce the example but when I try to run it with:

./combined-opt -i examples/phase_field-mechanics/hex_grain_growth_2D_eldrforce.i  Problem/type=DumpObjectsProblem Problem/dump_path=Modules/Combined/Master/grain_tracker

It is just saying Solve Converged and I don't see an output anywhere to look at. I think from the example that the dump path is asking for the name of the action that I want to look at right?

I do know that gr7 is not mapped to any grains in the domain initially, and that it only changes once a certain stress threshold is passed when I can see nuclei form but they don't affect the simulation it seems. See the attached images for initial and after significant stress. I have gr7 set to display the max and min value so you can see the nucleation is only getting it to about 0.048.

![Initial_gr7](https://user-images.githubusercontent.com/12799952/151028864-f8d4f372-cf9b-49d6-809b-0bc608c604a1.png)
![Nucleation_gr7_t0 6](https://user-images.githubusercontent.com/12799952/151028865-fb9e59ec-de3c-4da7-ba8d-e9aee52bac8f.png)

**Reply by laagesen** [reply]
I'm not sure that DiscreteNucleation is what's actually causing these nuclei to form. These nuclei don't look spherical and I think the DiscreteNucleationInserter can only insert spherical shapes. So here are a couple of suggestions:
-Try outputting the nucleation map to an AuxVariable using the DiscreteNucleationAux AuxKernel to see if the positions of these nuclei are in the same position as where the map is putting them
-Try running without applied stress. Unless the mesh displacement is significantly magnified it looks like there is a very high strain and probably very high elastic energy. It could be that the free energy elastic contribution is so large that it's becoming energetically favorable to form new nuclei in non-zero order parameters.

As far as the dump objects problem, I think if you leave the dump path blank it will just show the entire contents of the input file with actions substituted.

**Reply by jbair34** [reply]
I've output the nucleation map and it is in line with gr7. I think the reason they aren't circles is that the mesh doesn't adapt to gr7 right when the nuclei form so unless the spot is right on a grain boundary the mesh is pretty coarse. I am going to try increasing my nucleus size to see if maybe it is a mesh thing and the nuclei just aren't large enough to increase gr7 higher. I'll post how it goes later today.

![Nucleation_gr7](https://user-images.githubusercontent.com/12799952/151213624-ebb6f38d-71b9-42dc-9d0d-d4d73099c8ff.png)
![Nucleation_map_t0 4](https://user-images.githubusercontent.com/12799952/151213626-3606a7b3-cf74-43ba-8cad-4aab72bc8fb9.png)

**Reply by jbair34** [reply]
I ran it without stress. I also tried to have the nucleated spots have more refined meshes with a small amount of sucess. The max gr7 is still only getting to about 0.14. The nuc_map values corresponding with those locations is getting up to 1 but it isn't changing gr7 that much and it is also not changing the euler angles. Do you think I should try with a hold time more than the 0 recommended on mooseframework?

**Reply by laagesen** [reply]
I will ask @SudiptaBiswas to chime in here- she has done this successfully in UO2 with nucleation driven by dislocation density.

**Reply by SudiptaBiswas** [reply]
Hi Jake, Just from looking at the nucleation events, it appears to me that either the nuclei are too small or the mesh is too coarse. For the mesh, you can try the DiscreteNucleationMarker to refine the mesh around the nucleation location. https://mooseframework.inl.gov/modules/phase_field/Nucleation/DiscreteNucleation.html 
In some cases, I have seen nuclei disappear when the driving force is not enough for them to grow. It might be good to check if there is a critical nuclei size that you need for your physics.

**Reply by FHilty** [reply]
Hello All,

I am working with Jake on this project. We used the DiscreteNucleationMarker and that helped with our mesh, thank you for the tip Sudipta. We found that we needed to increase the nucleus_value in DiscreteNucleationForce such that the value the nuclei were taking on in gr7 was close to 1. Before it was only reaching 0.17, and thus not passing the reserve_op_threshold in the grain tracker so they weren't being transfered to a non-reserved op. Now we are having a new problem though, when the nucleus is created and meets the thresholds we set for it be included as a new grain the model crashes with an EBSD reader error:
```
*** ERROR ***
The following error occurred in the object "ebsd_reader", of type "EBSDReader".
Error! Index out of range in EBSDReader::indexFromIndex(), index: 21 size: 20
```
Our initial grain count is 20, so it seems that adding a new grain to the list is breaking the ebsd reader. We are using an EBSDMeshGenerator and our UserObjects block is:
```
[UserObjects]
  [ebsd_reader]
    type = EBSDReader
  []
  [ebsd]
    type = PolycrystalEBSD
    coloring_algorithm = jp
    ebsd_reader = ebsd_reader
    enable_var_coloring = true
  []
  [./grain_tracker]
    type = GrainTrackerElasticity
    threshold = 0.2
    compute_var_to_feature_map = true
    execute_on = 'initial timestep_begin'
    flood_entity_type = ELEMENTAL
    fill_method = symmetric9
    C_ijkl = '1.27e5 0.708e5 0.708e5 1.27e5 0.708e5 1.27e5 0.7355e5 0.7355e5 0.7355e5'
    euler_angle_provider = ebsd_reader
    polycrystal_ic_uo = ebsd
    reserve_op = 1
    reserve_op_threshold = 0.95
    outputs = exodus
  [../]
  #Nucleation user UserObjects
  [./inserter]
    type = DiscreteNucleationInserter
    hold_time = 0.1
    probability = P
    seed = 12346
    radius = 10
  [../]
  [./map]
    type = DiscreteNucleationMap
    int_width = 3
    periodic = gr7
    inserter = inserter
  [../]
[]
```
The EBSD reader seems to load all the information without any problems, it just fails when a nucleation event causes a new grain with the grain tracker. Any suggestions on how to resolve this?

Floyd

**Reply by jbair34** [reply]
Our non-EBSD simulation is also still having issues. Though the nuclei are getting up to 1 and remapping to new OPs, it doesn't look like they are showing as new grains yet. The euler angles seem to be changing to 0 in a few spots but that is the only change. Is it possible to make the euler angle a random value in nuclei?


![eulermap](https://user-images.githubusercontent.com/12799952/151708393-d65dbd43-fbb9-4610-945c-9e4e9068045a.png)
![gr6](https://user-images.githubusercontent.com/12799952/151708413-447befa4-1dee-4807-8037-8e02cd0ef27d.png)
![bnds](https://user-images.githubusercontent.com/12799952/151708421-f9c6b0b0-d80e-4864-a471-3548a985f7a5.png)

**Reply by SudiptaBiswas** [reply]
Hi Jake, 
What do you mean when you say 'Though the nuclei are getting up to 1 and remapping to new OPs, it doesn't look like they are showing as new grains yet'? Does it mean that the GrainTracker isn't remapping them onto new grain? If so, make sure you have `remap_grains = true`. Also, check that the threshold value is appropriate for the new grain/interface. I see that you have new nuclei formed within existing grains, which could indicate a split grain. Turn on the `tolerate_failure=true` and see if that helps with remapping. The EBSD size issue could be related to the remapping issue as well.

**Reply by ykvishal** [reply]
I don't know any solution to your problem, but I am curious about bnds  and gr plots. Are you using some modified version of [BndsCalcAux](https://mooseframework.inl.gov/moose/source/auxkernels/BndsCalcAux.html) as the max value is more than 1.  Similar trend is present for gr plot. For normal grain growth problem, both plots will have 1 as a max value. That is why I am curious.

**Reply by jbair34** [reply]
Hi Sudipta,
I am sorry I was misreading one of the console outputs that indicates the number of grains on each OP. They are showing up as new grains I am just dumb and was looking at the wrong spot on my console output. I am still concerned about the euler angles all going to 0 instead of random values though in new grains. Once we have enough grains I think that will just make them all merge right? So we do have remapping grains in the hexagonal starting point with the total number of grains increasing and I think with some fine tuning we will be able to get that one going but the EBSD one still will not increase the total number of grains without errors. Thanks again for your help and sorry I misread my own output yesterday!

Hi Vishal, 
We are using the normal bnds and gr parameters. The overly large values seems to be from my settings in nucleus value for the DiscreteNucleationForce kernel. I was trying to make sure they remapped to new grains and overdid it a little bit so I will be playing with that value a bit to optimize. But it seems like the nucleus_value needs to be at least about 13 or so for our case for any remapping to occur. In the images I posted I had it set to 17 which was giving bnds max at around 1.9 and gr max around 1.1. It seems to be a finicky parameter, I'm not getting consistent max OP values in my nuclei between runs.

**Reply by FHilty** [reply]
I decided to move my question to its own discussion, it seems divergent to Jakes question.
#20198

**Reply by SudiptaBiswas** [reply]
I am glad the issue is somewhat sorted. Regarding the EulerAngles, you will have to have an additional object that assigns random EulerAngles to the new grains, or for the EBSD case, it can be read from the file (might necessitate some addition to `GrainTracher`, I don't think it's done by default). For my simulations, I did not assign Euler angles to grains (it was more like an isotropic model), but I was still able to capture recrystallization.

**Reply by YXY101676** [reply]
Hi, I was wondering what your file does for GR7 as a variable in the Reaction kernel？

**Comment by YXY101676** [comment]
Hi, I want to know how you implemented nucleation in polycrystalline structure. I tried to combine the two files you mentioned, but there was an error that the variable could not be found.Can you teach me how to solve this problem?Thanks

**Reply by GiudGiud** [reply]
hello

please create your own post and include the input files and the error
you can reference this post for context of course

Guillaume
